### DISCUSSION TITLE
Serializing and deserializing an existing C++ structs from a third party library

### DISCUSSION THREAD

**Discussion by matusnovak** [capnproto/capnproto#1857] [Q&A] [answered]
I am curious if capnproto supports the following:

Let's say I have an already defined C++ struct that comes from a library. For example, the `glm::vec3` from https://glm.g-truc.net/0.9.2/api/a00236.html

Such 3D vector struct is defined as the following in C++:

```cpp
// Very simplified
namespace glm {
  struct vec3 {
    float x;
    float y;
    float z;
  };
}
```

I would like to create a capnproto schema that uses this already existing C++ struct in a message.

```
struct Object {
  name @0 :Text;
  position @1 :glm::vec3;
}
```

I could define `glm::vec3` inside the schema, but then I will end up with two `vec3` implementations, one from the third party library, and one created from the schema+generator. Moreover, that would require a conversion from schema-defined `vec3` C++ struct into the `glm::vec3` C++ struct after the deserialization happens, or before, when serialization happens. I would like to avoid that.

Is such thing possible?

I understand that other libraries might be better suited for this kind of serialization, such as MessagePack. I am exploring my options.

**Comment by kentonv** [comment, marked as answer]
No, sorry, Cap'n Proto is not intended to decode existing C-layout structs. Cap'n Proto defines its own specific layout, which has similarities to how C structs are laid out but is not intended to allow you to model arbitrary C structs.

**Reply by matusnovak** [reply]
That makes sense. Thank you for the confirmation.
